src/dsl/static.jl

```diff
@@ -66,184 +66,172 @@ function parse_julia_expr!(stmts, bindings, name::Symbol, var::Symbol,
-        builder, $(esc(gen_fn_or_dist)), inputs=[$(map(esc, inputs)...)], addr=$(QuoteNode(addr)),
-        name=$(QuoteNode(name)))))
-    true
+        builder, $(esc(gen_fn_or_dist)), inputs=[$(map(esc, inputs)...)],
+        addr=$(QuoteNode(addr)), name=$(QuoteNode(name)))))
+    # Return the name of the newly created node
+    return name
 end
 
-function parse_trainable_param!(stmts::Vector{Expr}, bindings, line::Expr)
-    if (line.head == :macrocall
-            && line.args[1] == STATIC_DSL_PARAM
-            && length(line.args) == 3
-            && isa(line.args[2], LineNumberNode))
-        (name::Symbol, typ) = parse_lhs(line.args[3])
-        if haskey(bindings, name)
-            static_dsl_syntax_error(addr_expr, "Symbol $name already bound")
-        end
-        node = gensym()
-        bindings[name] = node
-        push!(stmts, :($(esc(node)) = add_trainable_param_node!(
-            builder, $(QuoteNode(name)), typ=$(QuoteNode(typ)))))
-        true
-    else
-        return false
+function parse_trainable_param!(stmts::Vector{Expr}, bindings, expr::Expr)
+    (name::Symbol, typ) = parse_typed_var(expr)
+    if haskey(bindings, name)
+        static_dsl_syntax_error(expr, "Symbol $name already bound")
     end
+    node = gensym(name)
+    bindings[name] = node
+    push!(stmts, :($(esc(node)) = add_trainable_param_node!(
+        builder, $(QuoteNode(name)), typ=$(QuoteNode(typ)))))
+    true
 end
 
-function parse_trace_line!(stmts::Vector{Expr}, bindings, line::Expr)
-    if line.head == :(=)
-        @assert length(line.args) == 2
-        (lhs, rhs) = line.args
-        (name::Symbol, typ) = parse_lhs(lhs)
-        parse_trace_expr!(stmts, bindings, name, rhs, typ)
+function parse_return!(stmts::Vector{Expr}, bindings, expr)
+    if isa(expr, Symbol)
+        if !haskey(bindings, expr)
+            error("Tried to return $expr, which is not a locally bound variable")
+        end
+        node = bindings[expr]
     else
-        name = gensym()
-        typ = QuoteNode(Any)
-        parse_trace_expr!(stmts, bindings, name, line, typ)
+        name, typ = gensym("return"), QuoteNode(Any)
+        node = parse_julia_expr!(stmts, bindings, name, expr, typ)
+        bindings[name] = node
     end
+    push!(stmts, :(set_return_node!(builder, $(esc(node)))))
+    return Expr(:return, expr)
 end
 
-# return foo (must be a symbol) or return @trace(..)
-function parse_return!(stmts::Vector{Expr}, bindings, line::Expr)
-    if line.head != :return
-        return false
-    end
-    if isa(line.args[1], Expr) && line.args[1].head == :macrocall
-        var = gensym()
-        typ = QuoteNode(Any)
-        if !parse_trace_expr!(stmts, bindings, var, line.args[1], typ)
-            return false # the right-hand-side is a macro but not a valid `@trace` expression
-        end
-    elseif isa(line.args[1], Symbol)
-        var = line.args[1]
+function parse_expr!(stmts, bindings, expr)
```

Comment:
@ztangent It might be good to add a comment emphasizing that this function returns a new expression. Or maybe rename the function to `parse_and_simplify_expr!`.

Reply:
Makes sense! I'm happy with that, or maybe `process_expr!` to make clear that it's also modifying the expression in place? (@alex-lew did something similar with the `dist` DSL, and named the function `process_node`.)